### PR TITLE
feat: add user logout card

### PR DESCRIPTION
**Pull request by hperl**
This adds a user logout card that will be used for logging out users in the [OIDC front channel logout](https://openid.net/specs/openid-connect-frontchannel-1_0.html) flow, as specified in the [docs](https://www.ory.sh/docs/hydra/guides/logout).

Fixes https://github.com/ory-corp/cloud/issues/4173
Part of https://github.com/ory-corp/cloud/issues/4155

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [x] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation within the code base (if appropriate).

### CODE REVIEW

File: src/react-components/checkbox.tsx [hunk 1]
```diff
@@ -26,7 +26,7 @@ export const Checkbox = ({
   dataTestid,
   ...props
 }: CheckboxProps): JSX.Element => {
-  const id = Math.random().toString(36).substring(2)
+  const id = props.id ?? Math.random().toString(36).substring(2)
```
Comment (hperl) [inline src/react-components/checkbox.tsx:29]: This way things still work when we pass an id (as we do for the "remember me" checkbox).

File: src/react-components/ory/user-logout-card.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { expect, test } from "@playwright/experimental-ct-react"
```
Comment (hperl) [inline src/react-components/ory/user-logout-card.spec.tsx:1]: Playwright component tests are just so nice :heart_eyes:

Comment (mszekiel) [reply src/react-components/ory/user-logout-card.spec.tsx:1]: ngl, I love them

File: src/react-components/ory/user-logout-card.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+              type="submit"
+              id="accept"
+              value="Yes"
+              name="submit"
+              variant="semibold"
+              header="Yes"
+            />
+          </div>
+        </div>
+      </form>
```
Comment (Benehiko) [inline src/react-components/ory/user-logout-card.tsx:62]: Would we always manually be passing down the `challenge` and `csrftoken` to this component? Couldn't we just use the flow data, or if there isn't any, at least just a common type from the sdk?

Maybe we could rather extend this helper component? 
https://github.com/ory/elements/blob/main/src/react-components/ory/helpers/user-auth-form.tsx

The `UserAuthForm` already adds csrf and other hidden fields based on the flow data.

File: src/react-components/checkbox.tsx (hunk 1 above)
Comment (Benehiko) [reply src/react-components/checkbox.tsx:29]: makes sense :+1:

File: src/react-components/ory/user-logout-card.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  const [request] = await Promise.all([
+    component.page().waitForRequest("https://example.com/logout"),
+    component.locator("button", { hasText: /yes/i }).click(),
+  ])
+
+  expect(request.method()).toBe("POST")
+  const data = request.postDataJSON()
+  expect(data).toHaveProperty("_csrf", "CSRF token")
+  expect(data).toHaveProperty("challenge", "logout challenge")
+  expect(data).toHaveProperty("submit", "Yes")
```
Comment (Benehiko) [inline src/react-components/ory/user-logout-card.spec.tsx:27]: We also have this helper for testing `UserAuthPage` which you could pass along a flow object to check if the fields exist through a fixture.

So you have a fixture which is just json data stored [here](https://github.com/ory/elements/blob/main/src/test/Fixtures.ts) and then you have the `UserAuthPage` which has some methods to extract and test the fixtures against the actual rendered page [here](https://github.com/ory/elements/blob/main/src/test/AuthPage.ts).

Comment (Benehiko) [reply src/react-components/ory/user-logout-card.spec.tsx:27]: This of course assumes you are using flow data.

I think we should have these helpers in one place since they are packaged into the `@ory/elements-test` package. https://github.com/ory/elements/blob/main/packages/test/package.json

This package will be used as a tool by examples in Ory as well as, hopefully, our customers for tests. 
The idea is that we write tests once, have common selectors and common flows packaged which can be extended by each individual test case for edge cases. 

Here is a preview of an idea for common login tests 
https://github.com/ory/elements/blob/13551747b388932302390d46a1b38ae22caef02f/src/test/login/index.ts

Used by this react example 
https://github.com/ory/elements/blob/13551747b388932302390d46a1b38ae22caef02f/examples/react-spa/e2e/login.spec.ts

**Review by Benehiko — CHANGES_REQUESTED** [review]
To me this PR looks solid already, but I have some questions regarding how we expect the `UserLogoutCard` to be used. 

I also have some ideas on how we could improve this PR so that there is less work in future when it comes to testing.

File: src/react-components/ory/user-logout-card.tsx (hunk 3 above)
Comment (hperl) [reply src/react-components/ory/user-logout-card.tsx:62]: This card is not part of the Kratos flows, but implements the UI for the Hydra/OAuth 2 logout flow. The card will be initialized in the hosted UI here: https://github.com/ory-corp/cloud/pull/4176/files#diff-53a791f9f8393dc7f05d73059e3f2f4380ee68316b834ed244a57905d63c87f6R43-R49, but this is not part of any SDK, so we don't have a common type here. I just used the same parameters as the hydra-login-consent-node here: https://github.com/ory/hydra-login-consent-node/blob/master/views/logout.pug

**Review by Benehiko — APPROVED** [review]

**Comment by Benehiko** [comment]
As discussed, since this isn't part of the Kratos flows and is a custom implementation inside the Account Experience, we don't need to over-complicate the testing setup here. Also customization isn't currently necessary since it is only used inside Account Experience.

File: src/react-components/ory/user-logout-card.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import { gridStyle } from "../../theme"
+import { Button } from "../button"
+import { Card } from "../card"
+import { Typography } from "../typography"
+
+import { OAuth2Client, OAuth2ConsentRequest } from "@ory/client"
+
+import { Checkbox } from "../checkbox"
+import { Divider } from "../divider"
```
Comment (CaptainStandby) [inline src/react-components/ory/user-logout-card.tsx:9]: This is unused and can be removed

File: src/react-components/ory/user-logout-card.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      challenge="logout challenge"
+    />,
+  )
+})
+
+test("submits when user clicks 'Yes'", async () => {
+  const [request] = await Promise.all([
+    component.page().waitForRequest("https://example.com/logout"),
+    component.locator("button", { hasText: /yes/i }).click(),
+  ])
```
Comment (CaptainStandby) [inline src/react-components/ory/user-logout-card.spec.tsx:21]: This can be racy (see [here](https://www.notion.so/Playwright-E2E-Testing-Ory-Network-a88cba00aa3c4e3982bfa302e78be006?pvs=4#1f7b786a20ca48fb8c432c59f267cc53) and [here](https://playwright.dev/docs/api/class-page#page-wait-for-request))
You should first create the promise, then trigger the action and then await the promise.
```suggestion
  const requestPromise = component.page().waitForRequest("https://example.com/logout")
  await component.locator("button", { hasText: /yes/i }).click()
  const request = await requestPromise
```

File: src/react-components/ory/user-logout-card.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  expect(data).toHaveProperty("_csrf", "CSRF token")
+  expect(data).toHaveProperty("challenge", "logout challenge")
+  expect(data).toHaveProperty("submit", "Yes")
+})
+
+test("submits when user clicks 'No'", async () => {
+  const [request] = await Promise.all([
+    component.page().waitForRequest("https://example.com/logout"),
+    component.locator("button", { hasText: /no/i }).click(),
+  ])
```
Comment (CaptainStandby) [inline src/react-components/ory/user-logout-card.spec.tsx:34]: This can be racy (see [here](https://www.notion.so/Playwright-E2E-Testing-Ory-Network-a88cba00aa3c4e3982bfa302e78be006?pvs=4#1f7b786a20ca48fb8c432c59f267cc53) and [here](https://playwright.dev/docs/api/class-page#page-wait-for-request))
You should first create the promise, then trigger the action and then await the promise.
```suggestion
  const requestPromise = component.page().waitForRequest("https://example.com/logout")
  await component.locator("button", { hasText: /no/i }).click()
  const request = await requestPromise
```

**Review by CaptainStandby — CHANGES_REQUESTED** [review]
Looks fantastic. Only had two suggestions on writing more stable interactions in Playwright.

File: src/react-components/ory/user-logout-card.spec.tsx (hunk 6 above)
Comment (hperl) [reply src/react-components/ory/user-logout-card.spec.tsx:21]: My code is actually equivalent to the change you suggested.

Lets consider this: `await Promise.all([a, b])`, where `a` and `b` are promises. Then during the array construction, `a` will always be called before `b`.

So, `await Promise.all([a, b])` is equivalent to `const c = [a, b]; Promise.all(c)`, which is doing the right thing: Setting up the listener for the navigation first, then triggering the click event, then awaiting that both finish.

The only way I can see this race is if we were already on the destined URL. Otherwise I find this pattern more concise than what you suggested. WDYT?

Comment (hperl) [reply src/react-components/ory/user-logout-card.spec.tsx:21]: Hmm strange, in the bundled Playwright doc the recommendation is still to use `Promise.all`, but in the current version this was changed to the version you suggested. But I still can't see why this would race.

Comment (CaptainStandby) [reply src/react-components/ory/user-logout-card.spec.tsx:21]: I'm not sure, the order is always guaranteed like that. We actually had cases where this was causing flaky behaviour (though this was primarily with `waitForNaviagation` which is in itself problematic).
You are probably right that it is no problem in this case. The proposed pattern is guaranteed to work correctly though.

@jonas-jonas do you have some more insights into this?

Comment (CaptainStandby) [reply src/react-components/ory/user-logout-card.spec.tsx:21]: [here](https://github.com/microsoft/playwright/discussions/20779#discussioncomment-4942948) is a reason for the change in the documentation. They say there is no functional difference, but the newer pattern has fewer pitfalls for users.

Comment (hperl) [reply src/react-components/ory/user-logout-card.spec.tsx:21]: Ah, nice find. Then I'm happy, because my mental model of JS is still intact, and I'll still accept your suggestion because it has fewer pitfalls (which, in a language littered with pitfalls, is a strong argument).

**Review by CaptainStandby — APPROVED** [review]
LGTM 🎉
